Add PFNs4BO sampler

HideakiImamura opened the pull request:
## Contributor Agreements

Please read the [contributor agreements](https://github.com/optuna/optunahub-registry/blob/main/CONTRIBUTING.md#contributor-agreements) and if you agree, please click the checkbox below.

- [x] I agree to the contributor agreements.

## Motivation
<!-- Describe your motivation why you will submit this PR. This is useful for reviewers to understand the context of PR. -->
PFNs4BO was recently proposed by Samuel Müller, Matthias Feurer, Noah Hollmann, and Frank Hutter in https://arxiv.org/abs/2305.17535. It is highly interesting algorithm and can be applied to any BO problem. This PR aims to add the method in the OptunaHub.

## Description of the changes
<!-- Describe the changes in this PR. -->
- Add PFNs4BO sampler

HideakiImamura commented:
@y0z Could you review this PR?

HideakiImamura commented:
I conducted benchmark experiments of PFNs4BO against the random search. PFNs4BO looks promising.

![compare2-pfns4bo-vs-random-1](https://github.com/optuna/optunahub-registry/assets/38826298/d2325994-ac46-4ef6-b18a-cda2bc3cc07a)

This figure stands for the performance of PFNs4BO against the random search. The horizontal axis represents the amount of  budgets, which is equal to the number of trials. The vertical axis represents the p-value of the Mann–Whitney U test, in which the null hypothesis is that the PFNs4BO loses the random search. Therefore, don't get me wrong, but the lower is better.

The benchmark consists of 70 problems. The 8 problems are from HPO tabular benchmarks in https://github.com/automl/HPOBench (original 4 problems, and log-converted 4 problems). The 62 problems are several synthetic functions in https://github.com/sigopt/evalset (Not just use it, but I re-implement them).

SamuelGabriel commented on package/samplers/pfns4bo/sampler.py:
```diff
@@ -0,0 +1,274 @@
+        if isinstance(prior, torch.nn.Module):
+            trained_model = prior
+        elif prior == "vanilla gp":
+            _, _, trained_model, _ = train(**get_vanilla_gp_config(self._device))
```
I am not sure what your application is exactly. But training a PFN like the pre-trained ones takes around 1 day *with a GPU*. Thus, I don't think it is realistic HPO users want to do this, but they probably just want to use a pre-trained model. Either the ones I provided, or if you like to train better ones, then these.

HideakiImamura replied:
Sounds  very good. I completely agree with you. Since we should provide the pre-trained model for users, I will change the default value of the prior to use that provided in [here](https://github.com/automl/PFNs/blob/main/models_diff/prior_diff_real_checkpoint_n_0_epoch_42.cpkt).

SamuelGabriel replied:
No these: https://github.com/automl/PFNs4BO/tree/main/pfns4bo/final_models

SamuelGabriel commented on package/samplers/pfns4bo/sampler.py:
```diff
@@ -0,0 +1,274 @@
+        (
+            internal_search_space,
+            normalized_params,
+        ) = gp_search_space.get_search_space_and_normalized_params(trials, search_space)
```
The normalization of the HPs should be to [0,1] interval. Not sure, if that is what this function does. Just as a note :)

HideakiImamura replied:
Thanks for the comment. Yes, this search space transformation allow us to transform each variable into [0,1] interval.

SamuelGabriel commented on package/samplers/pfns4bo/sampler.py:
```diff
@@ -0,0 +1,274 @@
+                pre_sample_size=self._pre_sample_size,
+                device=self._device,
+                rand_sample_func=rand_sample_func,
+                apply_power_transform=False,
```
for us, power_transform was important to get really good scores. Maybe you can make it a hyper-parameter at least.

HideakiImamura replied:
Thank you for your input. I change the default value of it True.

SamuelGabriel commented on package/samplers/pfns4bo/sampler.py:
```diff
@@ -0,0 +1,274 @@
+        known_y = torch.from_numpy(standarized_score_vals).to(torch.float32).to(self._device)
+
+        with torch.enable_grad():
+            _, x_options, eis, _, _ = optimize_acq_w_lbfgs(
```
You can also pass an arg `acq_function` here, to specify to use another acquisition function from the ones specified in https://github.com/automl/PFNs4BO/blob/486e1ad36cb2d8e1ad933486a829e64395161b85/pfns4bo/scripts/acquisition_functions.py#L267 (some of these might not have gradients, though, but at least PI should also work)

HideakiImamura replied:
Thank you for the comment. It would be great for users to specify the type of acquisition functions. I will add the argument of `acquisition_function_type` to specify the type.

SamuelGabriel reviewed (commented):
To me this looks generally good. I think using power transform would improve performance further and you could also have something in place to use random search after 100 or 200 trials, as PFNs4BO becomes worse then.

Other than that: I did not double check all hyperparameters, but if you use the ones in the code (which are the same as in the paper), you should be fine.

HideakiImamura commented:
@SamuelGabriel Thank you for providing several important comments. I addressed all of your suggestions. Thank you vary much.

> using power transform would improve performance further

I will use `apply_power_transform = True` in the acquisition function optimization. Thanks for the information.

> you could also have something in place to use random search after 100 or 200 trials, as PFNs4BO becomes worse then.

Sounds very good. That type of algorithm change is easily possible in Optuna. Users can change the algorithm as follows. I will add that instruction in the docstring and README.

```python
sampler1 = PFNsBOSampler()
study = optuna.create_study(sampler=sampler)

study.optimize(objective, n_trials=200)

sampler2 = RandomSampler()
study.sampler = sampler2
study.optimize(objective, n_trials=100)
```


> Other than that: I did not double check all hyperparameters, but if you use the ones in the code (which are the same as in the paper), you should be fine.

I used completely same hyperparameters! Thank you for the information.

HideakiImamura commented:
@y0z Could you review this PR?

y0z commented on package/samplers/pfns4bo/README.md:
```diff
@@ -0,0 +1,67 @@
+---
+author: "HideakiImamura"
```
Don't need white space? (i.e., Hideaki Imamura)

y0z commented on package/samplers/pfns4bo/README.md:
```diff
@@ -0,0 +1,67 @@
+author: "HideakiImamura"
+title: "PFNs4BO sampler"
+description: "In-context learning for Bayesian optimization. This sampler uses Prior-data Fitted Networks (PFNs) as a surrogate model for Bayesian optimization."
+tags: ["sampler"]
```
Why not add tags to help users find your sampler?

y0z commented on package/samplers/pfns4bo/README.md:
````diff
@@ -0,0 +1,67 @@
+
+## Installation
+```bash
+pip install -r requirements.txt
````
The following will be more helpful because this allows users to install the dependencies by copy and paste. The URL becomes valid after merging this PR.

```suggestion
pip install -r https://raw.githubusercontent.com/optuna/optunahub-registry/main/package/samplers/pfns4bo/requirements.txt
```

HideakiImamura replied:
I think resolving dependencies might fail when the environment of users do not match the latest main branch. It would be great as it is.

y0z commented on package/samplers/pfns4bo/README.md:
````diff
@@ -0,0 +1,67 @@
+sampler = PFNs4BOSampler(prior=model)
+```
+
+The performance of PFNs4BO with the HEBO+ prior is maximized with the number of trials smaller than 100 or 200 in most cases. If you have a large number of trials, it is recommended to change the sampler to a random sampler or etc after a certain  number of trials.
````
```suggestion
The performance of PFNs4BO with the HEBO+ prior is maximized with the number of trials smaller than 100 or 200 in most cases. If you have a large number of trials, changing the sampler to another one (e.g., a random sampler) after a certain number of trials is recommended.
```

y0z commented on package/samplers/pfns4bo/README.md:
```diff
@@ -0,0 +1,67 @@
+
+### Reference
+
+Samuel Müller, Matthias Feurer, Noah Hollmann, and Frank Hutter. 2023. PFNs4BO: in-context learning for Bayesian optimization. In Proceedings of the 40th International Conference on Machine Learning (ICML'23), Vol. 202. JMLR.org, Article 1056, 25444–25470.
```
It seems helpful to add a link to [the paper](https://arxiv.org/abs/2305.17535).

y0z reviewed (commented):
I leave some review comments.

y0z reviewed (approved):
LGTM
